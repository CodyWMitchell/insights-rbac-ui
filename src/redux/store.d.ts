import { UserStore } from './reducers/user-reducer';

export type RBACStore = {
  userReducer: UserStore;
};

declare module 'react-redux' {
  export function useSelector<TState = RBACStore, TSelected = unknown>(
    selector: (state: TState) => TSelected,
    equalityFn?: (left: TSelected, right: TSelected) => boolean
  ): TSelected;
}
