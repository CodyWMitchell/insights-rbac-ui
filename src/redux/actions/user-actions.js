import * as ActionTypes from '../action-types';
import * as UserHelper from '../../helpers/user/user-helper';
import { createIntl, createIntlCache } from 'react-intl';
import messages from '../../Messages';
import providerMessages from '../../locales/data.json';
import { locale } from '../../AppEntry';

export const addUsers = (usersData) => ({
  type: ActionTypes.ADD_USERS,
  payload: UserHelper.addUsers(usersData).catch((err) => {
    const cache = createIntlCache();
    const intl = createIntl({ locale, messages: providerMessages }, cache);
    const error = err?.errors?.[0] || err;
    if (error.status === '400') {
      return {
        error: true,
      };
    }

    /**
     * Convert any other API error response to not crash notifications.
     * It has different format than other API requests.
     */
    throw {
      title: intl.formatMessage(messages.inviteUsersErrorTitle),
      message: intl.formatMessage(messages.inviteUsersErrorDescription),
      description: intl.formatMessage(messages.inviteUsersErrorDescription),
    };
  }),
});

export const updateUsers = (userList) => {
  const cache = createIntlCache();
  const intl = createIntl({ locale, messages: providerMessages }, cache);
  return {
    type: ActionTypes.UPDATE_USERS,
    payload: UserHelper.updateUsers(userList),
    meta: {
      notifications: {
        fulfilled: {
          variant: 'success',
          title: intl.formatMessage(messages.editUserSuccessTitle),
          dismissDelay: 8000,
          dismissable: true,
          description: intl.formatMessage(messages.editUserSuccessDescription),
        },
        rejected: {
          variant: 'danger',
          title: intl.formatMessage(messages.editUserErrorTitle),
          dismissDelay: 8000,
          dismissable: true,
          description: intl.formatMessage(messages.editUserErrorDescription),
        },
      },
    },
  };
};

export const fetchUsers = (apiProps) => ({
  type: ActionTypes.FETCH_USERS,
  payload: UserHelper.fetchUsers(apiProps),
});

export const updateUsersFilters = (filters) => ({
  type: ActionTypes.UPDATE_USERS_FILTERS,
  payload: filters,
});
