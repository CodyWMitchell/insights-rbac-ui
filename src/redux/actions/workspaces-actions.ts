import * as ActionTypes from '../action-types';
import * as WorkspacesHelper from '../../helpers/workspaces/workspaces-helper';

export const fetchWorkspaces = () => ({
  type: ActionTypes.FETCH_WORKSPACES,
  payload: WorkspacesHelper.getWorkspaces().catch((err) => {
    throw err;
  }),
});
