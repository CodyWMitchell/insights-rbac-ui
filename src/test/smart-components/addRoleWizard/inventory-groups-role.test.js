import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter } from 'react-router-dom';
import InventoryGroupsRole from '../../../smart-components/role/add-role/inventory-groups-role';
import { testInventoryGroups } from './inventory-groups-test-data';
import * as InventoryActions from '../../../redux/actions/inventory-actions';
import { FETCH_INVENTORY_GROUP } from '../../../redux/action-types';

const initialStateWithPermissions = {
  inventoryReducer: {
    isLoading: false,
    resourceTypes: {
      // this needs to be cleaned, just your previous setup did not match the desired structure
      'inventory:hosts:read': testInventoryGroups.reduce((acc, curr) => ({ ...acc, [curr.name]: { ...curr } }), {}),
      'inventory:hosts:write': testInventoryGroups.reduce((acc, curr) => ({ ...acc, [curr.name]: { ...curr } }), {}),
    },
    total: testInventoryGroups.length,
  },
};

jest.mock('@data-driven-forms/react-form-renderer/use-form-api', () => {
  return {
    __esModule: true,
    default: () => ({
      getState: jest.fn().mockReturnValue({
        values: {
          'add-permissions-table': [
            {
              uuid: 'inventory:hosts:read',
              requires: [],
            },
            {
              uuid: 'inventory:groups:write',
              requires: [],
            },
          ],
          'role-description': 'Be able to read and edit Account Staleness and Culling data.',
          'role-name': 'Account Staleness and Culling Administrator',
          'role-type': 'create',
          'role-uuid': '218ed86a-08f9-4d5d-aeac-7539eed882cd',
        },
      }),
      change: jest.fn(),
    }),
  };
});

jest.mock('@data-driven-forms/react-form-renderer/use-field-api', () => {
  return {
    __esModule: true,
    default: () => ({
      input: {
        value: ['inventory:hosts:read', 'inventory:groups:write'],
        name: 'inventory-groups-role',
        onChange: jest.fn(),
        onBlur: jest.fn(),
        onFocus: jest.fn(),
      },
    }),
  };
});

const renderComponent = (store) => {
  return render(
    <React.Fragment>
      <Provider store={store}>
        <BrowserRouter>
          <InventoryGroupsRole />
        </BrowserRouter>
      </Provider>
    </React.Fragment>
  );
};

describe('Inventory groups role', () => {
  let mockStore;

  const fetchInventoryGroupsSpy = jest.spyOn(InventoryActions, 'fetchInventoryGroups');

  beforeEach(() => {
    mockStore = configureStore();
  });

  afterEach(() => {
    fetchInventoryGroupsSpy.mockReset();
  });

  test('Add permissions to group renders without failing', async () => {
    fetchInventoryGroupsSpy.mockImplementationOnce(() => ({
      type: FETCH_INVENTORY_GROUP,
      payload: Promise.resolve({}),
    }));
    const store = mockStore(initialStateWithPermissions);

    const renderedResults = renderComponent(store);

    expect(renderedResults.getByText('Permissions')).toBeInTheDocument();
    expect()
  });

  test('Display available permission groups successfully', async () => {
    fetchInventoryGroupsSpy.mockImplementationOnce(() => ({
      type: FETCH_INVENTORY_GROUP,
      payload: Promise.resolve({}),
    }));

    const store = mockStore(initialStateWithPermissions);
    renderComponent(store);

    fireEvent.click(screen.getAllByLabelText('Options menu')[0]);

    expect(screen.getByText('fooBar')).toBeInTheDocument();
  });

  test('Selecting group for permission and Copy to all to other permissions successfully', async () => {
     fetchInventoryGroupsSpy.mockImplementationOnce(() => ({
      type: FETCH_INVENTORY_GROUP,
      payload: Promise.resolve({}),
    }));

    const store = mockStore(initialStateWithPermissions);
    const renderedResults = renderComponent(store);

    fireEvent.click(screen.getAllByLabelText('Options menu')[0]);
    expect(screen.getByText('fooBar')).toBeInTheDocument();
    fireEvent.click(screen.getByText('fooBar'));
    
    fireEvent.click(renderedResults.getByText('Copy to all'));
    expect(screen.getAllByLabelText('Options menu')[1]);
  });
});
