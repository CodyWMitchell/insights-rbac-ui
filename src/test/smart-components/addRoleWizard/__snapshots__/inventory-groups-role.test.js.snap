// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventory groups role Add permissions to group renders without failing 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-l-grid pf-m-gutter"
  >
    <div
      class="pf-v5-l-grid__item pf-m-6-col-on-md pf-m-3-col-on-lg rbac-m-hide-on-sm"
    >
      <h4
        class="rbac-bold-text pf-v5-u-mt-sm"
        data-ouia-component-id="OUIA-Generated-Text-1"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        Permissions
      </h4>
    </div>
    <div
      class="pf-v5-l-grid__item pf-m-6-col-on-md pf-m-9-col-on-lg rbac-m-hide-on-sm"
    >
      <h4
        class="rbac-bold-text pf-v5-u-mt-sm"
        data-ouia-component-id="OUIA-Generated-Text-2"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe="true"
        data-pf-content="true"
      >
        Group definition
      </h4>
    </div>
    <div
      class="pf-v5-l-grid"
    >
      <div
        class="pf-v5-l-grid__item pf-m-4-col-on-sm pf-m-3-col-on-md pf-m-3-col-on-lg"
      >
        <div
          class="pf-v5-c-form__group"
        >
          <div
            class="pf-v5-c-form__group-label"
          >
            <label
              class="pf-v5-c-form__label"
            >
              <span
                class="pf-v5-c-form__label-text"
              >
                hosts:read
              </span>
              <span
                aria-hidden="true"
                class="pf-v5-c-form__label-required"
              >
                 *
              </span>
            </label>
             
          </div>
          <div
            class="pf-v5-c-form__group-control"
          />
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-5-col-on-sm pf-m-6-col-on-md pf-m-7-col-on-lg"
      >
        <div
          style="display: contents;"
        >
          <div
            class="pf-v5-c-select rbac-m-resource-type-select"
            data-ouia-component-id="OUIA-Generated-Select-checkbox-1"
            data-ouia-component-type="PF5/Select"
            data-ouia-safe="true"
          >
            <div
              class="pf-v5-c-select__toggle pf-m-typeahead"
            >
              <div
                class="pf-v5-c-select__toggle-wrapper"
              >
                <span
                  class="pf-v5-c-select__toggle-text"
                >
                  Select groups
                </span>
              </div>
              <button
                aria-expanded="false"
                aria-label="Options menu"
                aria-labelledby="inventory:hosts:read pf-select-toggle-id-0"
                class="pf-v5-c-button pf-v5-c-select__toggle-button pf-m-plain"
                id="pf-select-toggle-id-0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg pf-v5-c-select__toggle-arrow"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 320 512"
                  width="1em"
                >
                  <path
                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-2-col-on-sm pf-m-4-col-on-md pf-m-2-col-on-lg"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-link"
          data-ouia-component-id="OUIA-Generated-Button-link-1"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          Copy to all
        </button>
      </div>
    </div>
    <div
      class="pf-v5-l-grid"
    >
      <div
        class="pf-v5-l-grid__item pf-m-4-col-on-sm pf-m-3-col-on-md pf-m-3-col-on-lg"
      >
        <div
          class="pf-v5-c-form__group"
        >
          <div
            class="pf-v5-c-form__group-label"
          >
            <label
              class="pf-v5-c-form__label"
            >
              <span
                class="pf-v5-c-form__label-text"
              >
                groups:write
              </span>
              <span
                aria-hidden="true"
                class="pf-v5-c-form__label-required"
              >
                 *
              </span>
            </label>
             
          </div>
          <div
            class="pf-v5-c-form__group-control"
          />
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-5-col-on-sm pf-m-6-col-on-md pf-m-7-col-on-lg"
      >
        <div
          style="display: contents;"
        >
          <div
            class="pf-v5-c-select rbac-m-resource-type-select"
            data-ouia-component-id="OUIA-Generated-Select-checkbox-2"
            data-ouia-component-type="PF5/Select"
            data-ouia-safe="true"
          >
            <div
              class="pf-v5-c-select__toggle pf-m-typeahead"
            >
              <div
                class="pf-v5-c-select__toggle-wrapper"
              >
                <span
                  class="pf-v5-c-select__toggle-text"
                >
                  Select groups
                </span>
              </div>
              <button
                aria-expanded="false"
                aria-label="Options menu"
                aria-labelledby="inventory:groups:write pf-select-toggle-id-1"
                class="pf-v5-c-button pf-v5-c-select__toggle-button pf-m-plain"
                id="pf-select-toggle-id-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg pf-v5-c-select__toggle-arrow"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 320 512"
                  width="1em"
                >
                  <path
                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-l-grid__item pf-m-2-col-on-sm pf-m-4-col-on-md pf-m-2-col-on-lg"
      />
    </div>
  </div>
</DocumentFragment>
`;
